Remove jQuery in Utils and Translation

sabas opened the pull request:
This pull request includes a

- [ ] Bug fix
- [ ] New feature
- [ ] Translation

The following changes were made

- I tried to remove jQuery from two files to see if I can help somehow
- I ran ```grunt test``` and all passed

If this is related to an existing ticket, include a link to it as well.

sabas committed:
Remove jQuery in Utils and Translation

sabas committed:
Remove jQuery in Utils and Translation

kevin-brown commented on src/js/select2/translation.js:
```diff
@@ -15,7 +14,7 @@ define([
 
   Translation.prototype.extend = function (translation) {
-    this.dict = $.extend({}, translation.all(), this.dict);
+    this.dict = Object.assign(translation.all(), this.dict);
```
Doesn't this modify the translation dictionary with the already defined translations? We were doing the extend with the empty object before to get around modifying the translations and to ensure it was always a new object.

sabas replied:
Ok sorry, I checked again the documentation and it can be done as in jquery, I add another commit

sabas committed:
Fix error

sabas committed:
Fix error

sabas committed:
Update translation.js

sabas committed:
Update translation.js

sabas commented:
#6196

jayaddison commented on src/js/select2/utils.js:
```diff
@@ -230,8 +228,14 @@ define([
+        paddingWidth += parseInt(computed.paddingRight, 10);
+  
+    return ((el.clientHeight - paddingHeight) < el.scrollHeight ||
+      (el.clientWidth - paddingWidth) < el.scrollWidth);
```
The `jQuery` approach seems more elegant and adaptable here - the API doesn't expose parsing of text values (such as `6px`) obtained from the computed style info.

Does this code follow a wide-recognized pattern for determining an element's inner dimensions, and if so is there a hyperlink to some documentation that explains it?

sabas replied:
@jayaddison actually I don't remember the rationale, probably I followed some ideas I found researching

jayaddison replied:
No problem.  I think it would be a significant benefit to remove the jQuery dependency; however I also remember why jQuery was originally popular -- to solve quirks and behaviours that can vary across browsers.  That means that I am sometimes a bit cautious about removing code that is expressed fairly clearly using jQuery -- it provides a single place to identify and solve those problems.

dtrucs commented on src/js/select2/utils.js:
```diff
@@ -230,8 +228,14 @@ define([
 
-    return ($el.innerHeight() < el.scrollHeight ||
-      $el.innerWidth() < el.scrollWidth);
+    var computed = window.getComputedStyle(el);
+    var paddingHeight = parseInt(computed.paddingTop, 10);
+        paddingHeight += parseInt(computed.paddingBottom, 10);
+    var paddingWidth = parseInt(computed.paddingLeft, 10);
+        paddingWidth += parseInt(computed.paddingRight, 10);
+  
+    return ((el.clientHeight - paddingHeight) < el.scrollHeight ||
+      (el.clientWidth - paddingWidth) < el.scrollWidth);
```
Why not just use the `width` and `height` from `getComputedStyle`?

```suggestion
    var computedEl = window.getComputedStyle(el);
 
    return (parseFloat(computedEl.height) < el.scrollHeight ||
      parseFloat(computedEl.width) < el.scrollWidth);
```

kevin-brown committed:
Simplify hasScroll and fix GetData dataset key lookup

Apply suggestion from review: use computedStyle.height/width directly
in hasScroll instead of manually summing padding values.

Fix GetData to convert attribute name format to dataset property key
format (e.g. 'foo-bar' -> 'fooBar') per the HTML spec, which fixes
the 'with nesting' data attributes test.
